File: ProjectCS/DisciplineList.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Collections.ObjectModel;

namespace SessionResultsCS
{
    // Класс для хранения списка дисциплин
    public class DisciplineList: IEnumerable<Discipline>
    {
        public const int DISCIPLINE_NOT_FOUND = -1;


        // Массив дисциплин
        private List<Discipline> _disciplines;


        public DisciplineList()
        {
            _disciplines = new List<Discipline>();
        }

        public DisciplineList(IEnumerable<Discipline> disciplines)
        {
            _disciplines = new List<Discipline>(disciplines);
        }


        public DisciplineList ShallowClone()
        {
            return new DisciplineList(_disciplines);
        }

        public DisciplineList DeepClone()
        {
            DisciplineList disciplineList = new DisciplineList();
            foreach (Discipline discipline in _disciplines)
            {
                disciplineList.AddDiscipline(discipline.Clone());
            }
            return disciplineList;
        }


        // Возвращает количество дисциплин в списке
        public int GetSize()
        {
            return _disciplines.Count;
        }

        // Возвращает дисциплину по указанному индексу
        public Discipline GetDisciplineAt(int index)
        {
            return _disciplines[index];
        }

        // Возвращает массив дисциплин
        public ReadOnlyCollection<Discipline> GetDisciplines()
        {
            return _disciplines.AsReadOnly();
        }

        // Находит дисциплину и возвращает её индекс. Если такой дисциплины нет в списке возвращает DISCIPLINE_NOT_FOUND
        public int FindDiscipline(Discipline discipline)
        {
            return _disciplines.IndexOf(discipline);
        }

        public IEnumerator<Discipline> GetEnumerator()
        {
            return _disciplines.GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }


        // Добавляет дисциплину в список и возвращает true при успешном добавлении, иначе false
        public bool AddDiscipline(Discipline discipline)
        {
            if (FindDiscipline(discipline) != DISCIPLINE_NOT_FOUND)
            {
                return false;
            }
            _disciplines.Add(discipline);
            return true;
        }

        // Убирает дисциплину из списка по индексу
        public void RemoveDisciplineAt(int index)
        {
            _disciplines.RemoveAt(index);
        }

        // Убирает дисциплину из списка и возвращает true при успешном удалении, иначе false
        public bool RemoveDiscipline(Discipline discipline)
        {
            int index = FindDiscipline(discipline);
            if (index == DISCIPLINE_NOT_FOUND)
            {
                return false;
            }
            _disciplines.RemoveAt(index);
            return true;
        }
    }
}

File: Project/Data/DisciplineList.cpp
#include "Data/DisciplineList.hpp"

#include <stdexcept>


DisciplineList::DisciplineList(): m_disciplines() {}

DisciplineList::DisciplineList(const std::vector<std::shared_ptr<Discipline>>& disciplines): m_disciplines(disciplines) {}

DisciplineList::DisciplineList(const DisciplineList& other): m_disciplines(other.m_disciplines) {}

DisciplineList DisciplineList::ShallowClone() const {
	return DisciplineList(m_disciplines);
}

DisciplineList DisciplineList::DeepClone() const {
	DisciplineList disciplineList;
	for (const std::shared_ptr<Discipline>& discipline : m_disciplines) {
		disciplineList.AddDiscipline(std::make_shared<Discipline>(*discipline));
	}
	return disciplineList;
}


int DisciplineList::GetSize() const {
	return m_disciplines.size();
}

std::shared_ptr<Discipline> DisciplineList::GetDisciplineAt(int index) const {
	return m_disciplines[index];
}

int DisciplineList::FindDiscipline(std::shared_ptr<Discipline> discipline) const {
	for (int i = 0; i < m_disciplines.size(); i++) {
		if (*m_disciplines[i] == *discipline) {
			return i;
		}
	}
	return DISCIPLINE_NOT_FOUND;
}

const std::vector<std::shared_ptr<Discipline>>& DisciplineList::GetDisciplines() const {
	return m_disciplines;
}


void DisciplineList::AddDiscipline(std::shared_ptr<Discipline> discipline) {
	if (FindDiscipline(discipline) != DISCIPLINE_NOT_FOUND) {
		throw std::invalid_argument("Added discipline is already in list");
	}
	m_disciplines.push_back(discipline);
}

void DisciplineList::RemoveDisciplineAt(int index) {
	m_disciplines.erase(m_disciplines.begin() + index);
}

void DisciplineList::RemoveDiscipline(std::shared_ptr<Discipline> discipline) {
	int index = FindDiscipline(discipline);
	if (index == DISCIPLINE_NOT_FOUND) {
		throw std::invalid_argument("Deleted discipline is not in list");
	}
	RemoveDisciplineAt(index);
}
